### ISSUE TITLE
Record syntax keys should be restricted to `string` values

### ISSUE THREAD

**Issue by CAD97** [nushell/nushell#10847]
### Related problem

This is sort of a "negative feature" request; potentially just a bug.

Record syntax (`{key: value}`) does a loose type coercion into `string`, accepting `int`, `float`, `string`, `binary`, or `datetime` by way of [`Value::as_string`](https://github.com/nushell/nushell/blob/7ac5a01e2f5020a1295697edab727cd62a5844b7/crates/nu-protocol/src/value/mod.rs#L328). Among other potentially surprising behavior, this means that `{0.0: value} == {0: value}` and `{0x[6b 65 79]: value} == {key: value}`.

*No* other syntax which accepts field names seems to have this behavior. Table syntax (`[[key]; [value]]`) requires the key to be a `string` (uses [`Expression::as_string`](https://github.com/nushell/nushell/blob/7ac5a01e2f5020a1295697edab727cd62a5844b7/crates/nu-protocol/src/ast/expression.rs#L110)), and `get <cell_path>` (has some of its own peculiarities[^cell-path] but) treats numbers as row indices, not column names.

[^cell-path]: e.g. `get 0x[6b 65 79]` is equivalent to `get '0x[6b 65 79]'`.

### Describe the solution you'd like

It's completely fair for record syntax to allow value keys; `{$key: $value}` is necessary functionality. But the type of the value should be restricted to just `string`, not doing this coercion.

### Describe alternatives you've considered

Just never write potentially confusing code?

### Additional context and details

| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.86.0                                   |
| branch             |                                          |
| commit_hash        | 5d8763ed1d14048ac9fc09eb21be36e6e622f0d7 |
| build_os           | windows-x86_64                           |
| build_target       | x86_64-pc-windows-msvc                   |
| rust_version       | rustc 1.71.1 (eb26296b5 2023-08-03)      |
| rust_channel       | 1.71.1-x86_64-pc-windows-msvc            |
| cargo_version      | cargo 1.71.1 (7f1d04c00 2023-07-29)      |
| build_time         | 2023-10-18 07:39:55 -05:00               |
| build_rust_channel | release                                  |
| allocator          | mimalloc                                 |
| features           | default, sqlite, trash, which, zip       |
| installed_plugins  |                                          |

**Comment by sholderbach** [comment]
I would consider this to be a parser bug.

**Comment by CAD97** [comment]
... [apparently](https://github.com/nushell/nushell/actions/runs/6687323421/job/18167834974?pr=10874#step:4:2996) there are examples taking advantage of integer literals as record keys. Are we still considering the ability to write this a bug?

**Comment by sholderbach** [comment]
Yeah, with how the type-directed parsing was envisioned this should probably be either an expression that directly evaluates to a string *or* a literal which as a word can also be reparsed as a string. 

so ok would be:
 - `("a" + "b")` -> evaluate to `"ab"`
 - `1` -> reparse to `"1"`
 - `1.0` -> reparse to `"1.0"`
 - bytes is probably a bit contentious (makes sense to evaluate direclty from incoming external data, some limitations around UTF-8. We shouldn't go through lossy conversion there)

cc @jntrnr things to consider for our parser fun to be more disciplined about

**Comment by CAD97** [comment]
To note, `^cmd` uses a similar "extended bareword" syntax, e.g. consider the difference in behavior between `^echo 0.0` and `run-external echo 0.0` — the former prints `0.0` but the latter `0`.

If record literals allow "extended barewords" for keys, it probably makes sense to extend this at least to table syntax keys, if not also cell-path.

The limits of cell-path / field access syntax being `.` delimited, along with the idea of maybe allowing `nested.field.path` field declarations (in nuon), mean that it might make sense to limit this to number literals without a `.`.

**Comment by sholderbach** [comment]
Good point about float-like barewords being incompatible with general cell-paths. If I recall our cellpath logic only lets you use an index for one level. This would make sense with this restriction, but it feels limiting and we should be able to relax that in the future.

#9082 would impose additional restrictions how you can pass strings with `.`
